File: node/src/session.ts
import {
    Bytes,
    PublicKey,
    PrivateKey,
    DHPrivateKey,
    SignaturePrivateKey,
    DHPublicKey,
    DHAlgorithm,
    CipherAlgorithm,
    DHKeyPair,
    SignaturePublicKey,
} from './crypto/types';
import * as proto from './proto';
import { SymmetricState } from './symmetricstate';
import { getDHAlgorithm, getCipherAlgorithm, getPublicAlgorithm } from './crypto';

const EMPTY = new Uint8Array(0);

function encodeUint32(n: number): Bytes {
    const buf = new Uint8Array(4);
    buf[0] = n & 0xff;
    buf[1] = (n >> 8) & 0xff;
    buf[2] = (n >> 16) & 0xff;
    buf[3] = (n >> 24) & 0xff;
    return buf;
}

export interface HandshakeResult {
    remotePublicKey?: PublicKey;
    peerAdditionalData?: Bytes;
}

export class Session {
    public send: (payload: proto.Payload) => void;
    public onReceive: (data: Bytes) => void = () => {};

    private handshakePromise: ((err: Error | null) => void) | null = null;
    private localState = new SymmetricState();
    private remoteState = new SymmetricState();

    private dhAlgorithm: DHAlgorithm | null = null;
    private ephemeralKeyPair: DHKeyPair | null = null;
    private remoteEphemeralPublicKey: DHPublicKey | null = null;

    private availableEphemeralKeyAlgorithms: proto.DHAlgorithm[] = [
        proto.DHAlgorithm.DHX25519,
        proto.DHAlgorithm.DHECC,
    ];
    private availableCipherAlgorithms: proto.CipherAlgorithm[] = [
        proto.CipherAlgorithm.CipherAesGcm,
    ];

    private remotePublicKey: PublicKey | null = null;
    private handshakeResult: HandshakeResult = {};
    private handshakeAdditionalData: Bytes | null = null;

    constructor(
        public readonly initiator: boolean,
        public readonly staticKeyPair: PrivateKey | null = null,
    ) {
        if (staticKeyPair?.isDHKey()) {
            const dhKey = staticKeyPair as DHPrivateKey;
            this.availableEphemeralKeyAlgorithms = [dhKey.getDHAlgorithmProto()];
        }
    }

    private getDHAlgorithm(candidate: proto.DHAlgorithm): DHAlgorithm {
        if (this.dhAlgorithm) {
            return this.dhAlgorithm;
        }
        if (this.staticKeyPair?.isDHKey()) {
            const dhKey = this.staticKeyPair as DHPrivateKey;
            return dhKey.getDHAlgorithm();
        }
        return getDHAlgorithm(candidate);
    }

    public handshake(additional?: Bytes | null): Promise<HandshakeResult> {
        if (additional) {
            this.handshakeAdditionalData = additional;
        }
        return new Promise<HandshakeResult>((resolve, reject) => {
            this.handshakePromise = (err) => {
                if (err) {
                    reject(err);
                } else {
                    if (this.remotePublicKey) {
                        this.handshakeResult.remotePublicKey = this.remotePublicKey;
                    }
                    resolve(this.handshakeResult);
                }
            }
            if (this.initiator) {
                this.sendHello(false).catch(err => {
                    if (this.handshakePromise) {
                        this.handshakePromise(err);
                    }
                });
            }
        });
    }

    public handleReceive(payload: proto.Payload): Promise<void> {
        switch (payload.payloadType) {
            case proto.PayloadType.PayloadHello:
                return this.handleHello(payload.data, false);
            case proto.PayloadType.PayloadHelloWithChangeAlgorithm:
                return this.handleHello(payload.data, true);
            case proto.PayloadType.PayloadEncryptedMessage:
                return this.handleEncryptedMessage(payload.data);
            default:
                return Promise.reject(new Error(`invalid payload type: ${payload.payloadType}`));
        }
    }

    public async write(data: Bytes): Promise<void> {
        const message = proto.EncryptedMessage.create({ data });
        const plaintext = proto.EncryptedMessage.encode(message).finish();
        const ciphertext = await this.localState.encryptWithAd(plaintext, EMPTY);

        const payload = proto.Payload.create({
            payloadType: proto.PayloadType.PayloadEncryptedMessage,
            data: ciphertext
        });
        this.send(payload);
    }

    private async handleHello(payload: Bytes, retry: boolean): Promise<void> {
        try {
            let handshakeFinish: boolean = false;
            let sendRetry: boolean = false;

            const hello = proto.Hello.decode(payload);
            const helloSigned = hello.signed!;
            const helloBytes = proto.HelloBytes.decode(payload);
            const helloSignedBytes = proto.HelloSignedBytes.decode(helloBytes.signed);

            // Check cipher algorithm compatibility
            if (!this.availableCipherAlgorithms.includes(helloSigned.cipherAlgorithm)) {
                if (retry) {
                    throw new Error('any cipher not supported');
                }

                this.availableCipherAlgorithms = this.availableCipherAlgorithms
                    .filter(algo => helloSigned.supportCipher.includes(algo));
                if (this.availableCipherAlgorithms.length === 0) {
                    throw new Error('any cipher not supported');
                }

                sendRetry = true;
            }

            let remotePublicKey: PublicKey | null = null;

            if (helloSigned.publicKey) {
                const publicKeyAlgorithm = getPublicAlgorithm(
                    helloSigned.publicKey.format,
                    helloSigned.dhAlsoPublicKey
                );
                remotePublicKey = await publicKeyAlgorithm.unmarshalPublicKey(helloSigned.publicKey.data);

                if (!helloSigned.dhAlsoPublicKey) {
                    const sigKey = remotePublicKey as SignaturePublicKey;
                    const verified = await sigKey.verify(helloBytes.signed, helloBytes.signature);
                    if (!verified) {
                        throw new Error('payload verification failed');
                    }
                    this.remotePublicKey = remotePublicKey;
                }
            }

            // Check DH algorithm compatibility
            if (!this.availableEphemeralKeyAlgorithms.includes(helloSigned.ephemeralKey!.algorithm)) {
                if (retry) {
                    throw new Error('any dh not supported');
                }

                this.availableEphemeralKeyAlgorithms = this.availableEphemeralKeyAlgorithms
                    .filter(algo => helloSigned.supportDh.includes(algo));
                if (this.availableEphemeralKeyAlgorithms.length === 0) {
                    throw new Error('any dh not supported');
                }

                sendRetry = true;
            }

            if (retry || sendRetry) {
                this.handshakeResult = {};
                this.localState = new SymmetricState();
                this.remoteState = new SymmetricState();
                this.ephemeralKeyPair = null;

                if (sendRetry) {
                    return this.sendHello(true);
                }
            }

            this.remoteState.mixHash(encodeUint32(hello.version));

            const cipherAlgorithm = getCipherAlgorithm(helloSigned.cipherAlgorithm);

            let dhAlgorithm: DHAlgorithm | null = null;
            if (remotePublicKey != null) {
                if (helloSigned.dhAlsoPublicKey) {
                    const dhKey = remotePublicKey as DHPublicKey;
                    dhAlgorithm = dhKey.getDHAlgorithm();
                    if (this.ephemeralKeyPair) {
                        const sharedKey = await this.ephemeralKeyPair.private.dh(dhKey);
                        this.remoteState.mixKey(cipherAlgorithm, sharedKey); // hello_02
                        this.localState.mixKey(cipherAlgorithm, sharedKey); // hello_02
                    }
                } else {
                    this.remoteState.mixHash(helloSignedBytes.publicKey); // hello_02
                }
                this.remotePublicKey = remotePublicKey;
            }
            if (dhAlgorithm == null) {
                dhAlgorithm = this.getDHAlgorithm(helloSigned.ephemeralKey!.algorithm);
            }
            this.dhAlgorithm = dhAlgorithm;

            this.remoteEphemeralPublicKey = await dhAlgorithm.unmarshalDHPublicKey(helloSigned.ephemeralKey!.data);

            if (this.ephemeralKeyPair == null && helloSigned.additional.length > 0) {
                this.handshakeResult.peerAdditionalData = await this.remoteState.mixHashAndDecrypt(helloSigned.additional);
            }

            if (this.staticKeyPair?.isDHKey()) {
                const localDHKey = this.staticKeyPair as DHPrivateKey;
                const sharedKey = await localDHKey.dh(this.remoteEphemeralPublicKey);
                this.localState.mixKey(cipherAlgorithm, sharedKey); // hello_02
                this.remoteState.mixKey(cipherAlgorithm, sharedKey); // hello_02
            }

            if (this.ephemeralKeyPair) {
                const sharedKey = await this.ephemeralKeyPair.private.dh(this.remoteEphemeralPublicKey);
                this.localState.mixKey(cipherAlgorithm, sharedKey);
                this.remoteState.mixKey(cipherAlgorithm, sharedKey);
                handshakeFinish = true;
            } else {
                this.remoteState.mixHash(helloSignedBytes.ephemeralKey);
            }

            if (this.ephemeralKeyPair != null && helloSigned.additional.length > 0) {
                this.handshakeResult.peerAdditionalData = await this.remoteState.mixHashAndDecrypt(helloSigned.additional);
            }

            if (helloBytes.signature.length > 0) {
                this.remoteState.mixHash(helloBytes.signature);
            }

            if (handshakeFinish) {
                if (this.handshakePromise) {
                    this.handshakePromise(null);
                }
            } else {
                await this.sendHello(false);
            }
        } catch (e: any) {
            if (this.handshakePromise) {
                this.handshakePromise(e);
                this.handshakePromise = null;
            }
        }
    }

    private async handleEncryptedMessage(payload: Bytes): Promise<void> {
        const plaintext = await this.remoteState.decryptWithAd(payload, EMPTY);
        const message = proto.EncryptedMessage.decode(plaintext);
        this.onReceive(message.data);
    }

    private async sendHello(changeAlgorithm: boolean): Promise<void> {
        let handshakeFinish = false;

        const dhAlgorithm = this.getDHAlgorithm(this.availableEphemeralKeyAlgorithms[0]);
        const cipherAlgorithm = getCipherAlgorithm(this.availableCipherAlgorithms[0]);

        const hello: proto.HelloBytes = {
            version: 1,
            signed: EMPTY,
            signature: EMPTY,
        };
        this.localState.mixHash(encodeUint32(hello.version));

        const helloSigned: proto.HelloSignedBytes = {
            supportDh: this.availableEphemeralKeyAlgorithms,
            supportCipher: this.availableCipherAlgorithms,
            cipherAlgorithm: cipherAlgorithm.getType(),
            additional: this.handshakeAdditionalData || EMPTY,
            dhAlsoPublicKey: false,
            publicKey: EMPTY,
            ephemeralKey: null as any,
        };

        if (this.staticKeyPair) {
            const publicKeyProto = await this.staticKeyPair.getPublic().marshalToProto();
            const publicKeyBytes = proto.PublicKey.encode(publicKeyProto).finish();
            helloSigned.publicKey = publicKeyBytes;

            if (this.staticKeyPair.isSignatureKey()) {
                this.localState.mixHash(publicKeyBytes);
            } else if (this.staticKeyPair.isDHKey()) {
                helloSigned.dhAlsoPublicKey = true;
            }
        }

        if (!this.ephemeralKeyPair) {
            this.ephemeralKeyPair = await dhAlgorithm.generate();

            if (this.remotePublicKey?.isDHKey()) {
                const remoteKey = this.remotePublicKey as DHPublicKey;
                const shared = await this.ephemeralKeyPair.private.dh(remoteKey);
                this.localState.mixKey(cipherAlgorithm, shared); // hello_02
                this.remoteState.mixKey(cipherAlgorithm, shared); // hello_02
            }
        }

        const ephemeralPublicKey: proto.DHPublicKey = {
            algorithm: dhAlgorithm.getType(),
            data: this.ephemeralKeyPair.public.marshal(),
        };
        const ephemeralKeyBytes = proto.DHPublicKey.encode(ephemeralPublicKey).finish();
        helloSigned.ephemeralKey = ephemeralKeyBytes;

        if (this.remoteEphemeralPublicKey) {
            const sharedKey = await this.ephemeralKeyPair.private.dh(this.remoteEphemeralPublicKey);
            this.localState.mixKey(cipherAlgorithm, sharedKey);
            this.remoteState.mixKey(cipherAlgorithm, sharedKey);
            handshakeFinish = true;
        } else {
            this.localState.mixHash(ephemeralKeyBytes);
        }

        if (this.handshakeAdditionalData && this.handshakeAdditionalData?.length > 0) {
            helloSigned.additional = await this.localState.encryptAndMixHash(
                this.handshakeAdditionalData,
                false
            );
        }

        const signedBytes = proto.HelloSignedBytes.encode(helloSigned).finish();
        hello.signed = signedBytes;

        if (this.staticKeyPair?.isSignatureKey()) {
            const signatureKey = this.staticKeyPair as SignaturePrivateKey;
            hello.signature = await signatureKey.sign(hello.signed);
            this.localState.mixHash(hello.signature);
        }

        if (handshakeFinish) {
            if (this.handshakePromise) {
                this.handshakePromise(null);
            }
        }

        const payload: proto.Payload = {
            payloadType: changeAlgorithm
                ? proto.PayloadType.PayloadHelloWithChangeAlgorithm
                : proto.PayloadType.PayloadHello,
            data: proto.HelloBytes.encode(hello).finish(),
        };
        this.send(payload);
    }
}

File: java/src/main/kotlin/kr/jclab/jscp/Session.kt
package kr.jclab.jscp

import com.google.protobuf.ByteString
import kr.jclab.jscp.crypto.*
import kr.jclab.jscp.payload.JcspPayload
import kr.jclab.jscp.state.SymmetricState
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.concurrent.CompletableFuture

class Session(
    private val initiator: Boolean,
    private val staticKeyPair: JscpPrivateKey? = null,
    private val send: (JcspPayload.Payload) -> Unit
) {
    var onReceive: ((ByteArray) -> Unit)? = null

    internal var symmetricStateFactory: (() -> SymmetricState)? = null

    internal lateinit var localState: SymmetricState
    internal lateinit var remoteState: SymmetricState

    private var dhAlgorithm: DHAlgorithm? = null
    private var ephemeralKeyPair: DHKeyPair? = null
    private var remoteEphemeralPublicKey: DHPublicKey? = null
    private var remotePublicKey: JscpPublicKey? = null

    private var handshakeResult = HandshakeResult()
    private var handshakePromise: ((Result<HandshakeResult>) -> Unit)? = null
    private var handshakeAdditionalData: ByteArray? = null

    private val availableEphemeralKeyAlgorithms = mutableListOf(
        JcspPayload.DHAlgorithm.DHX25519,
        JcspPayload.DHAlgorithm.DHECC
    )

    private val availableCipherAlgorithms = mutableListOf(
        JcspPayload.CipherAlgorithm.CipherAesGcm
    )

    init {
        if (staticKeyPair?.isDHKey() == true) {
            val dhKey = staticKeyPair as DHPrivateKey
            availableEphemeralKeyAlgorithms.clear()
            availableEphemeralKeyAlgorithms.add(dhKey.getDHAlgorithmProto())
        }
    }

    fun handshake(additional: ByteArray? = null): CompletableFuture<HandshakeResult> {
        val future = CompletableFuture<HandshakeResult>()

        localState = symmetricStateFactory?.invoke() ?: SymmetricState()
        remoteState = symmetricStateFactory?.invoke() ?: SymmetricState()

        handshakeAdditionalData = additional
        handshakePromise = { result ->
            result.fold(
                onSuccess = { future.complete(it) },
                onFailure = { future.completeExceptionally(it) }
            )
        }

        if (initiator) {
            try {
                sendHello(false)
            } catch (e: Exception) {
                future.completeExceptionally(e)
            }
        }

        return future
    }

    fun handleReceive(payload: JcspPayload.Payload) {
        try {
            when (payload.payloadType) {
                JcspPayload.PayloadType.PayloadHello ->
                    handleHello(payload.data.toByteArray(), false)
                JcspPayload.PayloadType.PayloadHelloWithChangeAlgorithm ->
                    handleHello(payload.data.toByteArray(), true)
                JcspPayload.PayloadType.PayloadEncryptedMessage ->
                    handleEncryptedMessage(payload.data.toByteArray())
                else -> throw IllegalArgumentException("Invalid payload type: ${payload.payloadType}")
            }
        } catch (e: Exception) {
            handshakePromise?.invoke(Result.failure(e))
        }
    }

    fun write(data: ByteArray) {
        try {
            val message = JcspPayload.EncryptedMessage.newBuilder()
                .setData(ByteString.copyFrom(data))
                .build()

            val plaintext = message.toByteArray()
            val ciphertext = localState.encryptWithAd(plaintext, ByteArray(0))

            val payload = JcspPayload.Payload.newBuilder()
                .setPayloadType(JcspPayload.PayloadType.PayloadEncryptedMessage)
                .setData(ByteString.copyFrom(ciphertext))
                .build()

            send(payload)
        } catch (e: Exception) {
            // Handle error
        }
    }

    private fun handleHello(payloadRaw: ByteArray, retry: Boolean) {
        var handshakeFinish = false
        var sendRetry = false

        val hello = JcspPayload.Hello.parseFrom(payloadRaw)
        val helloBytes = JcspPayload.HelloBytes.parseFrom(payloadRaw)
        val helloSignedBytes = JcspPayload.HelloSignedBytes.parseFrom(helloBytes.signed)

        var remotePublicKey: JscpPublicKey? = null
        if (hello.signed.hasPublicKey()) {
            val publicKeyAlgorithm = Registry.getPublicAlgorithm(
                hello.signed.publicKey.format,
                hello.signed.dhAlsoPublicKey
            )
            remotePublicKey = publicKeyAlgorithm.unmarshalPublicKey(
                hello.signed.publicKey.data.toByteArray()
            )

            if (!hello.signed.dhAlsoPublicKey) {
                val sigKey = remotePublicKey as SignaturePublicKey
                if (!sigKey.verify(helloBytes.signed.toByteArray(), helloBytes.signature.toByteArray())) {
                    throw IllegalStateException("Signature verification failed")
                }
            }
        }

        // Check cipher algorithm compatibility
        if (!availableCipherAlgorithms.contains(hello.signed.cipherAlgorithm)) {
            if (retry) {
                throw IllegalStateException("No cipher algorithm supported")
            }

            availableCipherAlgorithms.retainAll { hello.signed.supportCipherList.contains(it) }
            if (availableCipherAlgorithms.isEmpty()) {
                throw IllegalStateException("No cipher algorithm supported")
            }

            sendRetry = true
        }

        // Check DH algorithm compatibility
        if (!availableEphemeralKeyAlgorithms.contains(hello.signed.ephemeralKey.algorithm)) {
            if (retry) {
                throw IllegalStateException("No DH algorithm supported")
            }

            availableEphemeralKeyAlgorithms.retainAll { hello.signed.supportDhList.contains(it) }
            if (availableEphemeralKeyAlgorithms.isEmpty()) {
                throw IllegalStateException("No DH algorithm supported")
            }

            sendRetry = true
        }

        if (retry || sendRetry) {
            handshakeResult = HandshakeResult()
            localState = symmetricStateFactory?.invoke() ?: SymmetricState()
            remoteState = symmetricStateFactory?.invoke() ?: SymmetricState()
            ephemeralKeyPair = null

            if (sendRetry) {
                sendHello(true)
                return
            }
        }

        val cipherAlgorithm = Registry.getCipherAlgorithm(hello.signed.cipherAlgorithm)

        var dhAlgorithm: DHAlgorithm? = null

        remoteState.mixHash(encodeInt32(hello.version))

        if (remotePublicKey != null) {
            if (hello.signed.dhAlsoPublicKey) {
                dhAlgorithm = (remotePublicKey as DHPublicKey).getDHAlgorithm()
                ephemeralKeyPair?.let {
                    val sharedKey = it.private.dh(remotePublicKey)
                    remoteState.mixKey(cipherAlgorithm, sharedKey)
                    localState.mixKey(cipherAlgorithm, sharedKey)
                }
            } else {
                remoteState.mixHash(helloSignedBytes.publicKey.toByteArray())
            }
            this.remotePublicKey = remotePublicKey
            handshakeResult = handshakeResult.copy(
                remotePublicKey = remotePublicKey
            )
        }

        if (dhAlgorithm == null) {
            dhAlgorithm = getDHAlgorithm(hello.signed.ephemeralKey.algorithm)
        }
        this.dhAlgorithm = dhAlgorithm

        remoteEphemeralPublicKey = dhAlgorithm.unmarshalDHPublicKey(
            hello.signed.ephemeralKey.data.toByteArray()
        )

        if (ephemeralKeyPair == null && hello.signed.additional.size() > 0) {
            handshakeResult = handshakeResult.copy(
                peerAdditionalData = remoteState.mixHashAndDecrypt(hello.signed.additional.toByteArray())
            )
        }

        if (staticKeyPair?.isDHKey() == true) {
            val localDHKey = staticKeyPair as DHPrivateKey
            val sharedKey = localDHKey.dh(remoteEphemeralPublicKey!!)
            localState.mixKey(cipherAlgorithm, sharedKey)
            remoteState.mixKey(cipherAlgorithm, sharedKey)
        }

        if (ephemeralKeyPair != null) {
            val sharedKey = ephemeralKeyPair!!.private.dh(remoteEphemeralPublicKey!!)
            localState.mixKey(cipherAlgorithm, sharedKey)
            remoteState.mixKey(cipherAlgorithm, sharedKey)
            handshakeFinish = true
        } else {
            remoteState.mixHash(helloSignedBytes.ephemeralKey.toByteArray())
        }

        if (ephemeralKeyPair != null && hello.signed.additional.size() > 0) {
            handshakeResult = handshakeResult.copy(
                peerAdditionalData = remoteState.mixHashAndDecrypt(hello.signed.additional.toByteArray())
            )
        }

        if (helloBytes.signature.size() > 0) {
            remoteState.mixHash(helloBytes.signature.toByteArray())
        }

        if (handshakeFinish) {
            handshakePromise?.invoke(Result.success(handshakeResult))
        } else {
            sendHello(false)
        }
    }

    private fun handleEncryptedMessage(payloadRaw: ByteArray) {
        val plaintext = remoteState.decryptWithAd(payloadRaw, ByteArray(0))
        val message = JcspPayload.EncryptedMessage.parseFrom(plaintext)
        onReceive?.invoke(message.data.toByteArray())
    }

    private fun sendHello(changeAlgorithm: Boolean) {
        var handshakeFinish = false

        val dhAlgorithm = getDHAlgorithm(availableEphemeralKeyAlgorithms[0])
        val cipherAlgorithm = Registry.getCipherAlgorithm(availableCipherAlgorithms[0])

        val helloBuilder = JcspPayload.HelloBytes.newBuilder()
            .setVersion(1)

        localState.mixHash(encodeInt32(1))

        val helloSignedBuilder = JcspPayload.HelloSignedBytes.newBuilder()
            .addAllSupportDh(availableEphemeralKeyAlgorithms)
            .addAllSupportCipher(availableCipherAlgorithms)
            .setCipherAlgorithm(cipherAlgorithm.getType())

        if (staticKeyPair != null) {
            val publicKeyProto = staticKeyPair.getPublic().marshalToProto()
            val publicKeyBytes = publicKeyProto.toByteString()
            helloSignedBuilder.setPublicKey(publicKeyBytes)

            if (staticKeyPair.isSignatureKey()) {
                localState.mixHash(publicKeyBytes.toByteArray())
            } else if (staticKeyPair.isDHKey()) {
                helloSignedBuilder.setDhAlsoPublicKey(true)
            }
        }

        if (ephemeralKeyPair == null) {
            ephemeralKeyPair = dhAlgorithm.generate()

            if (remotePublicKey?.isDHKey() == true) {
                val remoteKey = remotePublicKey as DHPublicKey
                val shared = ephemeralKeyPair!!.private.dh(remoteKey)
                localState.mixKey(cipherAlgorithm, shared)
                remoteState.mixKey(cipherAlgorithm, shared)
            }
        }

        val ephemeralPublicKey = JcspPayload.DHPublicKey.newBuilder()
            .setAlgorithm(dhAlgorithm.getType())
            .setData(ByteString.copyFrom(ephemeralKeyPair!!.public.marshal()))
            .build()
        val ephemeralKeyBytes = ephemeralPublicKey.toByteString()
        helloSignedBuilder.setEphemeralKey(ephemeralKeyBytes)

        if (remoteEphemeralPublicKey != null) {
            val sharedKey = ephemeralKeyPair!!.private.dh(remoteEphemeralPublicKey!!)
            localState.mixKey(cipherAlgorithm, sharedKey)
            remoteState.mixKey(cipherAlgorithm, sharedKey)
            handshakeFinish = true
        } else {
            localState.mixHash(ephemeralKeyBytes.toByteArray())
        }

        if (handshakeAdditionalData != null && handshakeAdditionalData!!.isNotEmpty()) {
            val encrypted = localState.encryptAndMixHash(handshakeAdditionalData!!, false)
            helloSignedBuilder.setAdditional(ByteString.copyFrom(encrypted))
        }

        val signedBytes = helloSignedBuilder.build()
        val signedByteArray = signedBytes.toByteArray()
        helloBuilder.setSigned(ByteString.copyFrom(signedByteArray))

        if (staticKeyPair?.isSignatureKey() == true) {
            val signatureKey = staticKeyPair as SignaturePrivateKey
            val signature = signatureKey.sign(signedByteArray)
            helloBuilder.setSignature(ByteString.copyFrom(signature))
            localState.mixHash(signature)
        }

        if (handshakeFinish) {
            handshakePromise?.invoke(Result.success(handshakeResult))
        }

        val payloadType = if (changeAlgorithm) {
            JcspPayload.PayloadType.PayloadHelloWithChangeAlgorithm
        } else {
            JcspPayload.PayloadType.PayloadHello
        }

        val payload = JcspPayload.Payload.newBuilder()
            .setPayloadType(payloadType)
            .setData(helloBuilder.build().toByteString())
            .build()

        send(payload)
    }

    private fun getDHAlgorithm(candidate: JcspPayload.DHAlgorithm): DHAlgorithm {
        dhAlgorithm?.let { return it }

        if (staticKeyPair?.isDHKey() == true) {
            return (staticKeyPair as DHPrivateKey).getDHAlgorithm()
        }

        return Registry.getDHAlgorithm(candidate)
    }

    private fun encodeInt32(value: Int): ByteArray {
        return ByteBuffer.allocate(4)
            .order(ByteOrder.LITTLE_ENDIAN)
            .putInt(value)
            .array()
    }
}
